### PR TITLE
Fix autodetection of zip file format (including xlsx)

### PR DESCRIPTION
**Pull request by wm75**
addresses #8341 and adds integration and unit tests

### CODE REVIEW

File: lib/galaxy/util/compression_utils.py [hunk 1]
```diff
@@ -47,8 +47,10 @@ def get_fileobj_raw(filename, mode="r", compressed_formats=None):
         compressed_format = 'bz2'
     elif 'zip' in compressed_formats and zipfile.is_zipfile(filename):
         # Return fileobj for the first file in a zip file.
-        with zipfile.ZipFile(filename, mode) as zh:
-            fh = zh.open(zh.namelist()[0], mode)
+        # 'b' is not allowed in the ZipFile mode argument
```
Comment (mvdbeek) [inline lib/galaxy/util/compression_utils.py:50]:
```suggestion
        # 'b' is not allowed in the ZipFile mode argument
        # since files will always be opened in binary mode.
        # If text mode is required fh will be wrapped in
        # TextIOWrapper.
```

File: test/unit/test_compression_util.py [hunk 2]
```diff
@@ -2,20 +2,33 @@
 import tempfile
 import unittest
 
-from galaxy.util.compression_utils import CompressedFile
+from galaxy.util.compression_utils import CompressedFile, get_fileobj_raw
```
Comment (mvdbeek) [inline test/unit/test_compression_util.py]:
```suggestion
import six

from galaxy.util.compression_utils import (
    CompressedFile, 
    get_fileobj_raw,
)
```

This is just our import convention, and six is needed to assert we get the correct type from  `get_fileobj_raw `

File: test/unit/test_compression_util.py [hunk 3]
```diff
@@ -2,20 +2,33 @@
...
 
 
 class CompressionUtilTestCase(unittest.TestCase):
 
     def test_compression_safety(self):
-        self.assert_safety("test-data/unsafe.tar", False)
```
Comment (mvdbeek) [inline test/unit/test_compression_util.py]: Why did you remove this test ?

File: test/unit/test_compression_util.py [hunk 4]
```diff
@@ -26,3 +39,13 @@ def assert_safety(self, path, expected_to_be_safe):
                     CompressedFile(path).extract(temp_dir)
         finally:
             shutil.rmtree(temp_dir, ignore_errors=True)
+
+    def assert_format_detected(self, path, expected_fmt, allowed_fmts=None):
+        fmt1 = get_fileobj_raw(path, "r", allowed_fmts)[0]
```
Comment (mvdbeek) [inline test/unit/test_compression_util.py]: I'd do this in a for loop and also verify that we get the correct type:
```
for mode, expected_type in [('r', six.text_type), ('rb', six.binary_type), ('rt', six.text_type), ('U', six.text_type)]:
    fmt, fh = get_fileobj_raw(path, mode, allowed_fmts)
    assert fmt == expected_fmt
    assert isinstance(fh.read(1), expected_type)
```
I also like to avoid unittest's assertEqual (or unittest functionality in general), pytest has nice reporting for assertion errors.

**Review by mvdbeek — COMMENTED** [review]
Great, that looks correct to me, I'd just make the testing a bit more explicit.
Did you drop that one assert_safety test on purpose ?

File: test/unit/test_compression_util.py (hunk 3 above)
Comment (wm75) [reply test/unit/test_compression_util.py]: Oops, removed accidentally. Thanks for catching!

File: test/unit/test_compression_util.py (hunk 4 above)
Comment (wm75) [reply test/unit/test_compression_util.py]: Thanks for the nice idea of using the six types!
The only complication I had to deal with on top is that when the file format does not get autodetected, then, in text mode, the stream might not be decodable using utf-8. I decided to just skip the type check in this situation.

Comment (wm75) [reply test/unit/test_compression_util.py]: Ha, found a more elegant approach. It is possible to read 0 bytes/characters from the stream to learn whether it's binary or text. This avoids the encoding issue.

**Review by mvdbeek — APPROVED** [review]

File: lib/galaxy/util/compression_utils.py [hunk 5]
```diff
@@ -47,8 +47,13 @@ def get_fileobj_raw(filename, mode="r", compressed_formats=None):
...
+        # since it always opens files in binary mode.
+        # For emulating text mode, we will be returning the binary fh in a
+        # TextIOWrapper.
+        zf_mode = mode.replace('b', '')
+        with zipfile.ZipFile(filename, zf_mode) as zh:
+            fh = zh.open(zh.namelist()[0], zf_mode)
```
Comment (nsoranzo) [inline lib/galaxy/util/compression_utils.py:56]: @wm75 Will this work with a zipped binary file? I'd have thought to use `mode` instead of `zf_mode`.

Comment (mvdbeek) [reply lib/galaxy/util/compression_utils.py:56]: Yes, the mode is always binary, no matter what you select here. The unit tests assert this.
(indirectly though, by checking that that reading from the file handle returns `six.binary_type`)

Comment (wm75) [reply lib/galaxy/util/compression_utils.py:56]: @nsoranzo surprised me too, but the `zh.open` does not accept 'b' in mode either and always returns a binary stream.

Comment (wm75) [reply lib/galaxy/util/compression_utils.py:56]: @mvdbeek glad you suggested checking the stream type in the tests!

**Comment by martenson** [comment]
Thanks @wm75, this helps a lot!
